x/distribution/keeper/allocation.go

```diff
@@ -51,35 +51,63 @@ func (k Keeper) AllocateTokens(ctx context.Context, totalPreviousPower int64, bo
 	voteMultiplier := math.LegacyOneDec().Sub(communityTax)
 	feeMultiplier := feesCollected.MulDecTruncate(voteMultiplier)
 
-	// allocate tokens proportionally to voting power
-	//
-	// TODO: Consider parallelizing later
-	//
-	// Ref: https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r246276376
-	for _, vote := range bondedVotes {
+	var wg sync.WaitGroup
+	// channel for errors
+	errCh := make(chan error, len(bondedVotes))
+	// channel for remaining rewards
+	remainingCh := make(chan sdk.DecCoins, len(bondedVotes))
 
-		validator, err := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
-		if err != nil {
-			return err
-		}
+	for _, vote := range bondedVotes {
```

Comment:
nit: instead of using an error channel, consider instead using a https://pkg.go.dev/golang.org/x/sync/errgroup. It's much cleaner.

Also, IIRC, this is the first bit of concurrent code in the state machine. Are we sure this is safe?

Reply:
> Also, IIRC, this is the first bit of concurrent code in the state machine. Are we sure this is safe?

Discussed with Frojdi, and it may not be safe due to the concurrent writes (I thought it would have been, but no). Confirming with Matt and will revert that part if it isn't.

Reply:
we are writing to cachekv store which would be different than iavl concurrency. Maybe its safe?

Reply:
right, these land in memory first (a BTree in store v1, and probably a map in store v2).  store v2 is thread safe on get/set via a mutex, I think storev1 is the same.

so from the storage side I think concurrent processing is OK.

Reply:
Cool! That was my only doubt, the rest should be safe as it is basic calculation that does not depend on each other.

Reply:
Resolving :)